### PR TITLE
Fix incorrect target document calculation in template compilation

### PR DESCRIPTION
**Pull request by adrw**
Previously, if multiple installed documents had shadowed fields (ie.
`notificationId`), all of the documents would be added to any templates
that target one with the shadowed field. Thus, a SMS only template that
targets a SMS document with fields `sms_body` and `notificationId` will
end up having document targets include email and push documents which
shadow `notificationId` field.

This PR adds filtering and a test to only target documents that the template can satisfy based on the target Barber Signatures.

It also includes some misc code cleanup.

### CODE REVIEW

File: barber/src/main/kotlin/app/cash/barber/Barbershop.kt [hunk 1]
```diff
@@ -35,6 +35,11 @@ interface Barbershop {
   fun getWarnings(): List<String>
 
   interface Builder {
+    /**
+     * Configures this Barbershop so that instances of documentTemplate.templateToken will
+     * rendered by [documentTemplate] for its target locale.
+     */
+    @Deprecated("Use installDocumentTemplate(documentTemplate) instead")
```
Comment (lkerford) [inline barber/src/main/kotlin/app/cash/barber/Barbershop.kt:42]: Love it when people but the replacement variable in the deprecated annotation

File: barber/src/main/kotlin/app/cash/barber/models/CompiledDocumentTemplate.kt [hunk 2]
```diff
@@ -118,16 +119,23 @@ data class CompiledDocumentTemplate(
...
-              .firstOrNull { it is BarberField } as BarberField?
-          val mustache = mustacheFactoryProvider.get(barberField?.encoding)
-              .compile(StringReader(fieldValue), fieldValue)
-          val document = installedDocuments.row(signature)[fieldName]!!.document
-          documentTemplateFields.put(fieldName, document, mustache)
-        }
+        installedDocuments.column(fieldName).keys
+            .filter { signature ->
+              // Only add documents to table where the signature can be satisfied by the template
+              target_signatures.any {
```
Comment (lkerford) [inline barber/src/main/kotlin/app/cash/barber/models/CompiledDocumentTemplate.kt:125]: Should we notify people if a document doesn't satisfy the template?

**Review by lkerford — APPROVED** [review]

**Review by kshood — APPROVED** [review]

File: barber/src/main/kotlin/app/cash/barber/BarbershopBuilder.kt [hunk 3]
```diff
@@ -64,7 +64,7 @@ class BarbershopBuilder : Barbershop.Builder {
     installDocumentTemplate(documentTemplate.toProto())
   }
 
-  override fun installDocumentTemplate(documentTemplate: DocumentTemplate): Barbershop.Builder = apply {
```
Comment (ejao1) [inline barber/src/main/kotlin/app/cash/barber/BarbershopBuilder.kt:67]: is it considered more readable to leave out return types?

File: barber/src/main/kotlin/app/cash/barber/models/CompiledDocumentTemplate.kt (hunk 2 above)
Comment (adrw) [reply barber/src/main/kotlin/app/cash/barber/models/CompiledDocumentTemplate.kt:125]: It's less of something to warn or error on and more of a missed implementation detail that we didn't have sufficient test coverage to catch when the big PR went through

File: barber/src/main/kotlin/app/cash/barber/BarbershopBuilder.kt (hunk 3 above)
Comment (adrw) [reply barber/src/main/kotlin/app/cash/barber/BarbershopBuilder.kt:67]: IntelliJ puts them in implicitly and I was aiming for consistency with the other functions in the class

Comment (ejao1) [reply barber/src/main/kotlin/app/cash/barber/BarbershopBuilder.kt:67]: :+1:

**Review by ejao1 — APPROVED** [review]

File: barber/src/test/kotlin/app/cash/barber/BarbershopBuilderTest.kt [hunk 4]
```diff
@@ -161,6 +178,25 @@ class BarbershopBuilderTest {
     )
   }
 
+  @Test
+  fun `Barbershop does not include documents as targets where template does not satisfy all fields`() {
```
Comment (zcross) [inline barber/src/test/kotlin/app/cash/barber/BarbershopBuilderTest.kt:182]: 👍 very illustrative, nice

**Review by zcross — APPROVED** [review]
LGTM

**Pull request merged**